docs/proposals/authentication-filter.md

```diff
@@ -22,3 +22,1233 @@ This new filter should eventually expose all forms of authentication available t
+  - from: group=gateway.nginx.org, kind=AuthenticationFilter, namespace=<filter-namespace>
+  - to:   group="", kind=(Secret|ConfigMap), name=<target-name>
+- If no valid grant is found, the filter status should update the status to `Accepted=False` with `reason=RefNotPermitted` and a clear message. We should avoid rendering any NGINX configuration in this scenario.
+
+Example: Grant BasicAuth in app-ns to read a Secret in security-ns
+```yaml
+apiVersion: gateway.networking.k8s.io/v1
+kind: ReferenceGrant
+metadata:
+  name: allow-basic-auth-secret
+  namespace: security-ns # target namespace where the Secret lives
+spec:
+  from:
+  - group: gateway.nginx.org
+    kind: AuthenticationFilter
+    namespace: app-ns
+  to:
+  - group: ""   # core API group
+    kind: Secret
+    name: basic-auth-users
+```
+
+AuthenticationFilter referencing the cross-namespace Secret
+```yaml
+apiVersion: gateway.nginx.org/v1alpha1
+kind: AuthenticationFilter
+metadata:
+  name: basic-auth
+  namespace: app-ns
+spec:
+  type: Basic
+  basic:
+    secretRef:
+      namespace: security-ns
+      name: basic-auth-users
+      key: htpasswd
+    realm: "Restricted"
+```
+
+Example: Grant JWT file-based JWKS in keys-ns to filter in app-ns
+```yaml
+apiVersion: gateway.networking.k8s.io/v1
+kind: ReferenceGrant
+metadata:
+  name: allow-jwks-configmap
+  namespace: keys-ns
+spec:
+  from:
+  - group: gateway.nginx.org
+    kind: AuthenticationFilter
+    namespace: app-ns
+  to:
+  - group: ""    # core API group
+    kind: ConfigMap
+    name: jwt-keys
+```
+
+AuthenticationFilter referencing cross-namespace JWKS ConfigMap
+```yaml
+apiVersion: gateway.nginx.org/v1alpha1
+kind: AuthenticationFilter
+metadata:
+  name: jwt-auth
+  namespace: app-ns
+spec:
+  type: JWT
+  jwt:
+    mode: File
+    file:
+      configMapRef:
+        namespace: keys-ns
+        name: jwt-keys
+      mountPath: /etc/nginx/keys
+      fileName: jwks.json
+```
+
+### Remote JWKS
+
+Proxy cache TTL should be configurable and set to a resonable default, reducing periods of stale cached JWKs.
+
+### Key rotation
+
+Users sholud be advised to regularly rotate their JWKS keys in cases where they chose to reference a local JWKS via a `secrefRef` or `configMapRef`
+
+### Auth failure behaviour
+
+3xx response codes should not be allowed and AuthenticationFilter.onFailure must not support redirect targets. This is to prevent to prevent open-redirect abuse.
+
+401 and 403 should be the only allowable auth failure codes.
+
+### Auth failure default headers
+
+Below are a list of default defensive headers for authentication failure reponses.
+We may choose to include these headers by default for improved robustness in auth falure responses.
+
+```nginx
+add_header Content-Type "text/plain; charset=utf-8" always;
+add_header X-Content-Type-Options "nosniff" always;
+add_header Cache-Control "no-store" always;
+add_header Pragma "no-cache" always;
```

Comment:
I had to look this up, I had never seen it before. It's been superseded by Cache-Control, I don't think we would need it, it's legacy from HTTP/1.0